Activity tracking

mielliott opened the pull request:
New activities are now generated for parsing events and `wasInformedBy` statements document chains of cause and effect across activities. If all looks good, #33 and #41 should be closed.

Example output below. Note the connections between graph names and `wasInformedBy` statements.
```shell
$ preston track "http://ipt.vertnet.org:8080/ipt/rss.do"
$ preston ls | head -50
<https://preston.guoda.bio> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#SoftwareAgent> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<https://preston.guoda.bio> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Agent> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<https://preston.guoda.bio> <http://purl.org/dc/terms/description> "Preston is a software program that finds, archives and provides access to biodiversity datasets."@en <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<7fecd486-69bf-4bd0-8178-eeabe8891785> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Activity> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<7fecd486-69bf-4bd0-8178-eeabe8891785> <http://purl.org/dc/terms/description> "A crawl event that discovers biodiversity archives."@en <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<7fecd486-69bf-4bd0-8178-eeabe8891785> <http://www.w3.org/ns/prov#startedAtTime> "2020-03-21T20:08:59.544Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<7fecd486-69bf-4bd0-8178-eeabe8891785> <http://www.w3.org/ns/prov#wasStartedBy> <https://preston.guoda.bio> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<https://doi.org/10.5281/zenodo.1410543> <http://www.w3.org/ns/prov#usedBy> <7fecd486-69bf-4bd0-8178-eeabe8891785> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<https://doi.org/10.5281/zenodo.1410543> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://purl.org/dc/dcmitype/Software> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<https://doi.org/10.5281/zenodo.1410543> <http://purl.org/dc/terms/bibliographicCitation> "Jorrit Poelen, Icaro Alzuru, & Michael Elliott. 2019. Preston: a biodiversity dataset tracker (Version 0.0.1-SNAPSHOT) [Software]. Zenodo. http://doi.org/10.5281/zenodo.1410543"@en <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<0659a54f-b713-4f86-a917-5be166a14110> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Entity> <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<0659a54f-b713-4f86-a917-5be166a14110> <http://purl.org/dc/terms/description> "A biodiversity dataset graph archive."@en <7fecd486-69bf-4bd0-8178-eeabe8891785> .
<hash://sha256/e8e77bf19f00fbd28b0608b0c0507f642b2f622c737175819eae253fbb9dc1ac> <http://www.w3.org/ns/prov#wasGeneratedBy> <58412d96-482d-487b-9648-d5732781675d> <58412d96-482d-487b-9648-d5732781675d> .
<hash://sha256/e8e77bf19f00fbd28b0608b0c0507f642b2f622c737175819eae253fbb9dc1ac> <http://www.w3.org/ns/prov#qualifiedGeneration> <58412d96-482d-487b-9648-d5732781675d> <58412d96-482d-487b-9648-d5732781675d> .
<58412d96-482d-487b-9648-d5732781675d> <http://www.w3.org/ns/prov#generatedAtTime> "2020-03-21T20:09:00.712Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <58412d96-482d-487b-9648-d5732781675d> .
<58412d96-482d-487b-9648-d5732781675d> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Generation> <58412d96-482d-487b-9648-d5732781675d> .
<58412d96-482d-487b-9648-d5732781675d> <http://www.w3.org/ns/prov#wasInformedBy> <7fecd486-69bf-4bd0-8178-eeabe8891785> <58412d96-482d-487b-9648-d5732781675d> .
<58412d96-482d-487b-9648-d5732781675d> <http://www.w3.org/ns/prov#used> <http://ipt.vertnet.org:8080/ipt/rss.do> <58412d96-482d-487b-9648-d5732781675d> .
<http://ipt.vertnet.org:8080/ipt/rss.do> <http://purl.org/pav/hasVersion> <hash://sha256/e8e77bf19f00fbd28b0608b0c0507f642b2f622c737175819eae253fbb9dc1ac> <58412d96-482d-487b-9648-d5732781675d> .
<42ecab64-b575-46d3-8064-e590946064e9> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Activity> <42ecab64-b575-46d3-8064-e590946064e9> .
<42ecab64-b575-46d3-8064-e590946064e9> <http://www.w3.org/ns/prov#wasInformedBy> <58412d96-482d-487b-9648-d5732781675d> <42ecab64-b575-46d3-8064-e590946064e9> .
<hash://sha256/e8e77bf19f00fbd28b0608b0c0507f642b2f622c737175819eae253fbb9dc1ac> <http://www.w3.org/ns/prov#hadMember> <http://ipt.vertnet.org:8080/ipt/resource?id=dmns_bird_ggbn/v1.9> <42ecab64-b575-46d3-8064-e590946064e9> .
<http://ipt.vertnet.org:8080/ipt/resource?id=dmns_bird_ggbn/v1.9> <http://www.w3.org/ns/prov#hadMember> <http://ipt.vertnet.org:8080/ipt/eml.do?r=dmns_bird_ggbn> <42ecab64-b575-46d3-8064-e590946064e9> .
<http://ipt.vertnet.org:8080/ipt/eml.do?r=dmns_bird_ggbn> <http://purl.org/dc/elements/1.1/format> "application/eml" <42ecab64-b575-46d3-8064-e590946064e9> .
<hash://sha256/5fdf4bb4fe6a02fcc0d20a297446054164b3f585185654b1c6e9a6434b1e5d3b> <http://www.w3.org/ns/prov#wasGeneratedBy> <7989e193-47ed-4147-929a-e24ec3856047> <7989e193-47ed-4147-929a-e24ec3856047> .
<hash://sha256/5fdf4bb4fe6a02fcc0d20a297446054164b3f585185654b1c6e9a6434b1e5d3b> <http://www.w3.org/ns/prov#qualifiedGeneration> <7989e193-47ed-4147-929a-e24ec3856047> <7989e193-47ed-4147-929a-e24ec3856047> .
<7989e193-47ed-4147-929a-e24ec3856047> <http://www.w3.org/ns/prov#generatedAtTime> "2020-03-21T20:09:01.134Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <7989e193-47ed-4147-929a-e24ec3856047> .
<7989e193-47ed-4147-929a-e24ec3856047> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Generation> <7989e193-47ed-4147-929a-e24ec3856047> .
<7989e193-47ed-4147-929a-e24ec3856047> <http://www.w3.org/ns/prov#wasInformedBy> <42ecab64-b575-46d3-8064-e590946064e9> <7989e193-47ed-4147-929a-e24ec3856047> .
<7989e193-47ed-4147-929a-e24ec3856047> <http://www.w3.org/ns/prov#used> <http://ipt.vertnet.org:8080/ipt/eml.do?r=dmns_bird_ggbn> <7989e193-47ed-4147-929a-e24ec3856047> .
<http://ipt.vertnet.org:8080/ipt/eml.do?r=dmns_bird_ggbn> <http://purl.org/pav/hasVersion> <hash://sha256/5fdf4bb4fe6a02fcc0d20a297446054164b3f585185654b1c6e9a6434b1e5d3b> <7989e193-47ed-4147-929a-e24ec3856047> .
<http://ipt.vertnet.org:8080/ipt/resource?id=dmns_bird_ggbn/v1.9> <http://www.w3.org/ns/prov#hadMember> <http://ipt.vertnet.org:8080/ipt/archive.do?r=dmns_bird_ggbn> <42ecab64-b575-46d3-8064-e590946064e9> .
<http://ipt.vertnet.org:8080/ipt/archive.do?r=dmns_bird_ggbn> <http://purl.org/dc/elements/1.1/format> "application/dwca" <42ecab64-b575-46d3-8064-e590946064e9> .
<hash://sha256/5e91f618302e6d7516a3e3eeaa283400c693a55ad76dbb0f947cee3fc6749e57> <http://www.w3.org/ns/prov#wasGeneratedBy> <260734cc-e8f1-42ba-bc43-e87f5609cef9> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<hash://sha256/5e91f618302e6d7516a3e3eeaa283400c693a55ad76dbb0f947cee3fc6749e57> <http://www.w3.org/ns/prov#qualifiedGeneration> <260734cc-e8f1-42ba-bc43-e87f5609cef9> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<260734cc-e8f1-42ba-bc43-e87f5609cef9> <http://www.w3.org/ns/prov#generatedAtTime> "2020-03-21T20:09:02.048Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<260734cc-e8f1-42ba-bc43-e87f5609cef9> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Generation> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<260734cc-e8f1-42ba-bc43-e87f5609cef9> <http://www.w3.org/ns/prov#wasInformedBy> <42ecab64-b575-46d3-8064-e590946064e9> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<260734cc-e8f1-42ba-bc43-e87f5609cef9> <http://www.w3.org/ns/prov#used> <http://ipt.vertnet.org:8080/ipt/archive.do?r=dmns_bird_ggbn> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<http://ipt.vertnet.org:8080/ipt/archive.do?r=dmns_bird_ggbn> <http://purl.org/pav/hasVersion> <hash://sha256/5e91f618302e6d7516a3e3eeaa283400c693a55ad76dbb0f947cee3fc6749e57> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<hash://sha256/e8e77bf19f00fbd28b0608b0c0507f642b2f622c737175819eae253fbb9dc1ac> <http://www.w3.org/ns/prov#hadMember> <http://ipt.vertnet.org:8080/ipt/resource?id=msb_para_ggbn/v1.9> <42ecab64-b575-46d3-8064-e590946064e9> .
<http://ipt.vertnet.org:8080/ipt/resource?id=msb_para_ggbn/v1.9> <http://www.w3.org/ns/prov#hadMember> <http://ipt.vertnet.org:8080/ipt/eml.do?r=msb_para_ggbn> <42ecab64-b575-46d3-8064-e590946064e9> .
<http://ipt.vertnet.org:8080/ipt/eml.do?r=msb_para_ggbn> <http://purl.org/dc/elements/1.1/format> "application/eml" <42ecab64-b575-46d3-8064-e590946064e9> .
<hash://sha256/349a0937d43a545fbd21af43a61ff0677e44befb262ccfe9119b5dbdfc17aaff> <http://www.w3.org/ns/prov#wasGeneratedBy> <215a0533-dbdb-457d-be22-2f8c780c1fbe> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
<hash://sha256/349a0937d43a545fbd21af43a61ff0677e44befb262ccfe9119b5dbdfc17aaff> <http://www.w3.org/ns/prov#qualifiedGeneration> <215a0533-dbdb-457d-be22-2f8c780c1fbe> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
<215a0533-dbdb-457d-be22-2f8c780c1fbe> <http://www.w3.org/ns/prov#generatedAtTime> "2020-03-21T20:09:02.157Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
<215a0533-dbdb-457d-be22-2f8c780c1fbe> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/prov#Generation> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
<215a0533-dbdb-457d-be22-2f8c780c1fbe> <http://www.w3.org/ns/prov#wasInformedBy> <42ecab64-b575-46d3-8064-e590946064e9> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
<215a0533-dbdb-457d-be22-2f8c780c1fbe> <http://www.w3.org/ns/prov#used> <http://ipt.vertnet.org:8080/ipt/eml.do?r=msb_para_ggbn> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
<http://ipt.vertnet.org:8080/ipt/eml.do?r=msb_para_ggbn> <http://purl.org/pav/hasVersion> <hash://sha256/349a0937d43a545fbd21af43a61ff0677e44befb262ccfe9119b5dbdfc17aaff> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
```
```shell
$ preston ls | grep "wasInformedBy" | head
<58412d96-482d-487b-9648-d5732781675d> <http://www.w3.org/ns/prov#wasInformedBy> <7fecd486-69bf-4bd0-8178-eeabe8891785> <58412d96-482d-487b-9648-d5732781675d> .
<42ecab64-b575-46d3-8064-e590946064e9> <http://www.w3.org/ns/prov#wasInformedBy> <58412d96-482d-487b-9648-d5732781675d> <42ecab64-b575-46d3-8064-e590946064e9> .
<7989e193-47ed-4147-929a-e24ec3856047> <http://www.w3.org/ns/prov#wasInformedBy> <42ecab64-b575-46d3-8064-e590946064e9> <7989e193-47ed-4147-929a-e24ec3856047> .
<260734cc-e8f1-42ba-bc43-e87f5609cef9> <http://www.w3.org/ns/prov#wasInformedBy> <42ecab64-b575-46d3-8064-e590946064e9> <260734cc-e8f1-42ba-bc43-e87f5609cef9> .
<215a0533-dbdb-457d-be22-2f8c780c1fbe> <http://www.w3.org/ns/prov#wasInformedBy> <42ecab64-b575-46d3-8064-e590946064e9> <215a0533-dbdb-457d-be22-2f8c780c1fbe> .
```

jhpoelen commented on src/test/java/bio/guoda/preston/cmd/CmdUpdateTest.java:
```diff
@@ -66,7 +65,12 @@ public void doUpdate() throws IOException {
+        long informed = quads.stream()
+                .filter(x -> x.getPredicate().equals(WAS_INFORMED_BY))
+                .count();
+        assertThat(informed, is(unique - 1)); // Only the first activity should lack WAS_INFORMED_BY
```
suggest to refactor the assertion into a method that reflects the language in the comment (e.g., assertActivitiesWereInformed) . I try to avoid using comments in favor of descriptive method and property names.

mielliott replied:
Trying to put the assertion into its own method got a little messy since it relies on some variables earlier in the tests, so I just reworded variable names to make the assertion clearer (see 9a5d9c4217bf04d8c7be81b0df3ff118fb3fac2a).

jhpoelen commented on src/main/java/bio/guoda/preston/process/ActivityUtil.java:
```diff
@@ -48,9 +45,14 @@ private static void emitWithActivityName(Stream<Quad> quadStream, StatementEmitt
+        return newActivity;
+    }
+
+    public static void emitAsNamedActivity(Stream<Quad> quadStream, StatementEmitter emitter, Optional<BlankNodeOrIRI> parentActivity, BlankNodeOrIRI activityName) {
```
@mielliott nice descriptive method name! The name ````emitAsNamedActivity``` suggests that an activity is emitted. However, the statement is emitted as part of a activity. Would something like ```emitAsPartOfActivity``` reflect the inner workings of the method?

mielliott replied:
`emitWithActivityName` is the one that simply adds activity labels to provided quads, but `emitAsNamedActivity` is supposed to be an alternative to `emitAsNewActivity` to allow the activity ID to be specified. That is, `emitAsNamedActivity` should emit all of the activity statements, start to finish. Maybe renaming to `emitAsNewNamedActivity` would make sense?

jhpoelen replied:
```emitAsNewNamedActivity``` makes sense. Thanks for elaborating.

jhpoelen commented on src/test/java/bio/guoda/preston/process/RegistryReaderALATest.java:
```diff
@@ -72,11 +77,11 @@ public InputStream get(IRI key) throws IOException {
+        Assert.assertThat(nodes.size(), is(3982));
         Quad secondPage = nodes.get(nodes.size() - 3);
-        assertThat(secondPage.toString(), is("<https://collections.ala.org.au/ws/dataResource/dr8052> <http://purl.org/pav/createdBy> <https://ala.org.au> ."));
+        assertThat(secondPage, hasTriple(toStatement(toIRI("https://collections.ala.org.au/ws/dataResource/dr8052"), toIRI("http://purl.org/pav/createdBy"), toIRI("https://ala.org.au"))));
```
Nice! Thanks for replacing these strings with typed statement assertions.

jhpoelen commented:
@mielliott thanks for preparing the pull request! I left some review comments. Curious to hear your replies.

mielliott commented:
> 
> 
> @mielliott when using constants (i.e. any static final), please capitalize names and use snake case (e.g., EMIT_STARTED_AT). However, why was emitStartedAt introduced?

Yikes! Those little constants weren't really supposed to be permanent. Just an awkward way to bring up discussion about which statements should be included when emitting an activity. I've removed the code related to those constants and added to the discussion in #33.

Thanks for the feedback!
